swarm/src/lib.rs

```diff
@@ -135,29 +142,111 @@ pub type NegotiatedSubstream = Negotiated<Substream<StreamMuxerBox>>;
+        cause: ConnectionError<NodeHandlerWrapperError<THandleErr>>,
+    },
+    /// A new connection arrived on a listener and is now starting to handshake.
+    ///
+    /// A corresponding [`NewConnection`](SwarmEvent::NewConnection),
+    /// [`BannedPeer`](SwarmEvent::BannedPeer), or
+    /// [`IncomingConnectionError`](SwarmEvent::IncomingConnectionError) event will later be
+    /// generated for this connection.
+    IncomingConnection {
+        /// Local connection address.
+        /// This address has been earlier reported with a [`NewListenAddr`](SwarmEvent::NewListenAddr)
+        /// event.
+        local_addr: Multiaddr,
+        /// Address used to send back data to the remote.
+        send_back_addr: Multiaddr,
+    },
+    /// An error happened on a connection during its initial handshake.
+    ///
+    /// This can include, for example, an error during the handshake of the encryption layer, or
+    /// the connection unexpectedly closed.
+    IncomingConnectionError {
+        /// Local connection address.
+        /// This address has been earlier reported with a [`NewListenAddr`](SwarmEvent::NewListenAddr)
+        /// event.
+        local_addr: Multiaddr,
+        /// Address used to send back data to the remote.
+        send_back_addr: Multiaddr,
+        /// The error that happened.
+        error: PendingConnectionError<io::Error>,
+    },
+    /// We connected to a peer, but we immediately closed the connection because that peer is banned.
+    BannedPeer {
+        /// Identity of the banned peer.
+        peer_id: PeerId,
+        /// Endpoint of the connection that has been closed.
+        endpoint: ConnectedPoint,
+    },
+    /// Starting to try to reach the given peer.
+    ///
+    /// We are trying to connect to this peer until a [`NewConnection`](SwarmEvent::NewConnection)
+    /// event is reported, or a [`UnreachableAddr`](SwarmEvent::UnreachableAddr) event is reported
+    /// with `attempts_remaining` equal to 0.
+    StartConnect(PeerId),
     /// Tried to dial an address but it ended up being unreachaable.
     UnreachableAddr {
-        /// `PeerId` that we were trying to reach. `None` if we don't know in advance which peer
-        /// we were trying to reach.
-        peer_id: Option<PeerId>,
+        /// `PeerId` that we were trying to reach.
+        peer_id: PeerId,
         /// Address that we failed to reach.
         address: Multiaddr,
         /// Error that has been encountered.
-        error: Box<dyn error::Error + Send>,
+        error: PendingConnectionError<io::Error>,
+        /// Number of remaining connection attempts that are being tried for this peer.
+        attempts_remaining: u32,
+    },
+    /// Tried to dial an address but it ended up being unreachaable.
+    /// Contrary to `UnreachableAddr`, we don't know the identity of the peer that we were trying
+    /// to reach.
+    // TODO: This variant can be in an inconsistent state if `error == PeerIdMismatch`
```

Comment:
As per the other comment, this is not quite right, because there is only `PendingConnectionError::InvalidPeerId` which can also be triggered if the remote unexpectedly "identified" with the local peer ID.